src/main/java/com/amazon/opendistroforelasticsearch/sql/executor/AsyncRestExecutor.java

```diff
@@ -97,24 +102,56 @@ private boolean isRunningInTransportThread() {
     private void async(Client client, Map<String, String> params, QueryAction queryAction, RestChannel channel) {
         // Run given task in thread pool asynchronously
         client.threadPool().schedule(
-            new TimeValue(0L),
-            SQL_WORKER_THREAD_POOL_NAME,
             () -> {
                 try {
-                    executor.execute(client, params, queryAction, channel);
+                    doExecuteWithTimeMeasured(client, params, queryAction, channel);
                 } catch (IOException | SqlParseException e) {
-                    LOG.warn("[MCB] async task got an IO/SQL exception: {}", e.getMessage());
+                    LOG.warn("[{}] [MCB] async task got an IO/SQL exception: {}", requestId(queryAction), e.getMessage());
                     channel.sendResponse(new BytesRestResponse(RestStatus.INTERNAL_SERVER_ERROR, e.getMessage()));
                 } catch (IllegalStateException e) {
-                    LOG.warn("[MCB] async task got a runtime exception: {}", e.getMessage());
+                    LOG.warn("[{}] [MCB] async task got a runtime exception: {}", requestId(queryAction), e.getMessage());
                     channel.sendResponse(new BytesRestResponse(RestStatus.INSUFFICIENT_STORAGE, "Memory circuit is broken."));
                 } catch (Throwable t) {
-                    LOG.warn("[MCB] async task got an unknown throwable: {}", t.getMessage());
+                    LOG.warn("[{}] [MCB] async task got an unknown throwable: {}", requestId(queryAction), t.getMessage());
                     channel.sendResponse(new BytesRestResponse(RestStatus.INTERNAL_SERVER_ERROR, String.valueOf(t.getMessage())));
                 } finally {
                     BackOffRetryStrategy.releaseMem(executor);
                 }
-            });
+            },
+            new TimeValue(0L),
+            SQL_WORKER_THREAD_POOL_NAME);
+    }
+
+    private void sync(Client client, Map<String, String> params, QueryAction queryAction, RestChannel channel) {
+        try {
+            doExecuteWithTimeMeasured(client, params, queryAction, channel);
+        }
+        catch (Exception e) {
+            LOG.error(String.format("[%s] Failed during query execution in executor", requestId(queryAction)), e);
+            channel.sendResponse(new BytesRestResponse(SERVICE_UNAVAILABLE, new ErrorMessage(e, SERVICE_UNAVAILABLE.getStatus()).toString()));
+        }
+    }
+
+    /** Time the real execution of Executor and log slow query for troubleshooting */
+    private void doExecuteWithTimeMeasured(Client client,
+                                           Map<String, String> params,
+                                           QueryAction action,
+                                           RestChannel channel) throws Exception {
+        Instant startTime = Instant.now();
```

Comment:
Instant.now() can have pretty big errors depending on time drift/skew/last time of sync of system time/etc. Please use `System.nanoTime()` for anything where you need to measure a period and not get a specific point in time (https://docs.oracle.com/javase/8/docs/api/java/lang/System.html#nanoTime--).

Reply:
> Instant.now() can have pretty big errors depending on time drift/skew/last time of sync of system time/etc. Please use `System.nanoTime()` for anything where you need to measure a period and not get a specific point in time (https://docs.oracle.com/javase/8/docs/api/java/lang/System.html#nanoTime--).

Thanks for the suggestion! I wasn't aware of this. Could you pass me some reference for the problem you mentioned? Because we only care about the delta value returned from `Duration.between`, is it possible to be wrong too in some case?

Reply:
This one touches upon the problem a little bit: https://www.baeldung.com/java-measure-elapsed-time. Most of the sources say that resolution of nanoTime is better, but that's not the issue. The real difference is that `System.currentTimeMillis()` and `Instant.now()` (which relies on the former) both use wall-clock time from the JVM, which in its turn gets it from OS. Depending on what protocol the machine is using to synchronize the time, it can lag behind or go far ahead and then during sync get a huge change (either negative or positive). The `System.nanoTime()` is based on software-based counter (not related to wall clock), which only increases, and does not need to be synced with any other machine/server/clock, so the diffs of values from nanoTime will always be pretty accurate (I guess as long as you are using it on the same thread).

Reply:
> This one touches upon the problem a little bit: https://www.baeldung.com/java-measure-elapsed-time. Most of the sources say that resolution of nanoTime is better, but that's not the issue. The real difference is that `System.currentTimeMillis()` and `Instant.now()` (which relies on the former) both use wall-clock time from the JVM, which in its turn gets it from OS. Depending on what protocol the machine is using to synchronize the time, it can lag behind or go far ahead and then during sync get a huge change (either negative or positive). The `System.nanoTime()` is based on software-based counter (not related to wall clock), which only increases, and does not need to be synced with any other machine/server/clock, so the diffs of values from nanoTime will always be pretty accurate (I guess as long as you are using it on the same thread).

Thanks! Good to know the subtle difference. I think `Instant` and `Duration` is sufficient in our case. Because we are not benchmarking, we just need a rough elapsed time in second(s). `nanoTime()` may be more accurate in some edge case, though it's hassle to convert it to sec or millisec for logging or comparison. Hopefully this makes sense to you.

Reply:
It's just two lines that need to be changed:
#140 should be `final long startNanos = System.nanoTime();`
#145 should be `final Duration elapsed = Duration.ofNanos(System.nanoTime() - startNanos);

It's up to you if you want to keep it this way, not a blocker.

Reply:
> It's just two lines that need to be changed:
> #140 should be `final long startNanos = System.nanoTime();`
> #145 should be `final Duration elapsed = Duration.ofNanos(System.nanoTime() - startNanos);
> 
> It's up to you if you want to keep it this way, not a blocker.

Easier than I thought. Testing it. Will push very soon. Thanks!